src/service/position-observer-impl.js

```diff
@@ -0,0 +1,389 @@
+      this.scrollTimer_ = Date.now();
+      this.schedulePass_();
+      setTimeout(() => {
+        if (Date.now() - this.scrollTimer_ < 100) {
+          // viewport scroll in the last 100 ms, wait
+          return;
+        }
+        // assume scroll stops, if no scroll event in the last 100,
+        this.inScroll_ = false;
+      }, 100);
+    }));
+    this.unlisteners_.push(this.viewport_.onChanged(() => {
+      this.vsync_.measure(() => {
+        this.pass_(true);
+      });
+    }));
+  }
+
+  /* @override */
+  stopCallback() {
+    while(this.unlisteners_.length) {
+      const unlisten = this.unlisteners_.pop();
+      unlisten();
+    }
+  }
+
+  /** @param {boolean=} recursive */
+  schedulePass_(recursive) {
+    // TODO:
+    // P1: account for effective fidelity using this.effectiveFidelity
+    // P1[done]: use every vsyn measure() for HIGH
+    // P2[done]: use every 1 in 3 vsyn measure() for LOW
+    // P1[done]: only do this while user is scrolling.
+    // P1: also do passes for onResize,
+    // P2: do passes on onDomMutation (if available using MutationObserver)
+    if (!recursive && this.measure_) {
+      // call of schedulePass_ from viewport onScroll
+      // Do nothing if currently measure with calling schedulePass recursively
+      return;
+    }
+    this.measure_ = true;
+    if (!this.inScroll_) {
+      // not in scroll, do not need to measure
+      this.measure_ = false;
+      return;
+    }
+    this.vsync_.measure(() => {
+      this.pass_();
+      this.schedulePass_(true);
+    });
+  }
+
+  /**
+   * @param {boolean=} force
+   * @private
+  */
+  pass_(force) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      if (!force && entry.turn != 0) {
+        // Not ready for their turn yet.
+        entry.turn--;
+        continue;
+      }
+      // Reset entry.turn value.
+      if (!force) {
+        entry.turn = (entry.fidelity == PositionObserverFidelity.LOW) ?
+          LOW_FEDELITY_FRAME_COUNT : 0;
+      }
+
+      this.updateEntryPosition_(entry);
+    }
+  }
+
+  /** @override */
+  updateEntryPosition_(entry) {
+    // get layoutBoxes relative to doc.
+    let elementBox = this.viewport_.getLayoutRect(entry.element);
+    let viewportBox = this.viewport_.getRect();
+
+    // Adjust relative coordinate to viewport
+    // TODO(@zhouyx): We did one move when in #getLayoutRect.
+    // Optimize the code to avoid duplicate moveLayoutRect
+    elementBox = moveLayoutRect(elementBox,
+        -viewportBox.left, -viewportBox.top);
+    viewportBox = moveLayoutRect(viewportBox,
+        -viewportBox.left, -viewportBox.top);
+
+    // Return { positionRect: <LayoutRectDef>, viewportRect: <LayoutRectDef>}
+    entry.trigger(/** @type {PositionInViewportEntryDef}*/ ({
+      positionRect: elementBox,
+      viewportRect: viewportBox,
+    }));
+  }
+}
+
+export class InaboxPositionObserver extends AbstractPositionObserver {
```

Comment:
I think both implementations should be in different modules b/c they operate in vastly different environments: AMP host vs foreign page.